Output UnionType str rep with () instead of []

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69502

Differential Revision: [D32902781](https://our.internmc.facebook.com/intern/diff/D32902781)

tugsbayasgalan committed:
Output UnionType str rep with () instead of []

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32902781).

gmagogsfm commented on test/test_jit.py:
```diff
@@ -3635,6 +3635,19 @@ def check_subclass_warn(input: torch.LongTensor) -> torch.LongTensor:
+        # TS graph IR for Union should be annotated as Union()
+        FileCheck().check("Union(").check("int?").run(graph_rep)
+        # Serialized code for Union should be annotated as Union[]
+        FileCheck().check("Union[").check("Optional[int]").run(code_rep)
```
Indentation is weird?

tugsbayasgalan replied:
looks fine to me?

gmagogsfm commented on test/test_jit.py:
```diff
@@ -3635,6 +3635,19 @@ def check_subclass_warn(input: torch.LongTensor) -> torch.LongTensor:
+        # Serialized code for Union should be annotated as Union[]
+        FileCheck().check("Union[").check("Optional[int]").run(code_rep)
+        self.checkScript(test_func, (5, 6))
+
```
Maybe call `parse_ir` on `graph_rep` to verify round-trip-ability.

tugsbayasgalan replied:
sounds good

gmagogsfm commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1141,16 +1144,16 @@ std::string UnionType::unionStr(TypePrinter printer, bool is_annotation_str) con
 
 std::string UnionType::str() const {
-  return this->unionStr(nullptr, /*is_annotation_str=*/false);
+  return this->unionStr(nullptr, /*is_annotation_str=*/false, "(", ")");
```
You can probably use `is_annotation_str` to branch on which delimiter to use.

tugsbayasgalan replied:
sounds good

tugsbayasgalan committed:
Update on "Output UnionType str rep with () instead of []"


Differential Revision: [D32902781](https://our.internmc.facebook.com/intern/diff/D32902781)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32902781).

gmagogsfm commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1116,7 +1116,10 @@ std::string UnionType::unionStr(TypePrinter printer, bool is_annotation_str) con
+  std::string open_delimeter = is_annotation_str ? "[" : "(";
+  std::string close_delimeter = is_annotation_str ? "]" : ")";
+
+  ss << "Union" + open_delimeter;
```
```suggestion
  ss << "Union" << open_delimeter;
```
